finetrainers/trainer.py

```diff
@@ -0,0 +1,542 @@
+
+    def prepare_dataset(self) -> None:
+        logger.info("Initializing dataset and dataloader")
+
+        self.dataset = VideoDatasetWithResizing(
+            data_root=self.args.data_root,
+            caption_column=self.args.caption_column,
+            video_column=self.args.video_column,
+            resolution_buckets=self.args.video_resolution_buckets,
+            dataset_file=self.args.dataset_file,
+            id_token=self.args.id_token,
+        )
+        self.dataloader = torch.utils.data.DataLoader(
+            self.dataset,
+            batch_size=1,
+            sampler=BucketSampler(self.dataset, batch_size=self.args.batch_size, shuffle=True),
+            collate_fn=self.model_config.get("collate_fn"),
+            num_workers=self.args.dataloader_num_workers,
+            pin_memory=self.args.pin_memory,
+        )
+    
+    def prepare_trainable_parameters(self) -> None:
+        logger.info("Initializing trainable parameters")
+
+        # TODO(aryan): refactor later. for now only lora is supported
+        self.text_encoder.requires_grad_(False)
+        self.transformer.requires_grad_(False)
+        self.vae.requires_grad_(False)
+
+        # For mixed precision training we cast all non-trainable weights (vae, text_encoder and transformer) to half-precision
+        # as these weights are only used for inference, keeping weights in full precision is not required.
+        weight_dtype = torch.float32
+        if self.state.accelerator.mixed_precision == "fp16":
+            weight_dtype = torch.float16
+        elif self.state.accelerator.mixed_precision == "bf16":
+            weight_dtype = torch.bfloat16
+        
+        if torch.backends.mps.is_available() and weight_dtype == torch.bfloat16:
+            # due to pytorch#99272, MPS does not yet support bfloat16.
+            raise ValueError(
+                "Mixed precision training with bfloat16 is not supported on MPS. Please use fp16 (recommended) or fp32 instead."
+            )
+        
+        # TODO(aryan): handle torch dtype from accelerator vs model dtype
+        self.state.weight_dtype = weight_dtype
+        self.text_encoder.to(self.state.accelerator.device, dtype=weight_dtype)
+        self.transformer.to(self.state.accelerator.device, dtype=weight_dtype)
+        self.vae.to(self.state.accelerator.device, dtype=weight_dtype)
+
+        if self.args.gradient_checkpointing:
+            self.transformer.enable_gradient_checkpointing()
+        
+        transformer_lora_config = LoraConfig(
+            r=self.args.rank,
+            lora_alpha=self.args.lora_alpha,
+            init_lora_weights=True,
+            target_modules=self.args.target_modules,
+        )
+        self.transformer.add_adapter(transformer_lora_config)
+
+        # TODO: refactor
+        # create custom saving & loading hooks so that `accelerator.save_state(...)` serializes in a nice format
+        def save_model_hook(models, weights, output_dir):
```

Comment:
DeepSpeed is missed in this case.

Reply:
Yes, I've not added any DeepSpeed support yet. This was very fragile and I want to take it up another PR so that any model can be used easily without an if-else hell